Always use discriminant in ZeroTrie Serde; use serialize_bytes consistently

sffc opened the pull request:
Fixes #4562

There are diffs; good to get this into 1.5

robertbastian commented on utils/zerotrie/src/serde.rs:
```diff
@@ -132,7 +139,10 @@ where
         } else {
             let bytes = self.as_bytes();
-            bytes.serialize(serializer)
+            let mut all_in_one_vec = Vec::with_capacity(bytes.len() + 1);
+            all_in_one_vec.push(Self::FLAGS);
+            all_in_one_vec.extend(bytes);
+            all_in_one_vec.serialize(serializer)
```
this can be done without allocating a vec, also below

```suggestion
            let mut seq = serializer.serialize_seq(Some(bytes.len() + 1))?;
            seq.serialize_element(Self::FLAGS);
            for b in self.bytes {
                seq.serialize_element(b)?;
            }
            seq.end()
```

sffc replied:
Actually I should be explicitly using `serialize_bytes`. I don't want `serialize_seq`

sffc replied:
I think postcard represents serialize_bytes and serialize_seq the same way, but I think that's not a guarantee, and on the other end we definitely want deserialize_borrowed_bytes

robertbastian replied:
I'm not sure `serialize_bytes` is better than `serialize_seq` if you have to reallocate the whole vector in order to use it.

sffc replied:
The [contract of `serialize_bytes`](https://docs.rs/serde/latest/serde/trait.Serializer.html#tymethod.serialize_bytes) says:

> Enables serializers to serialize byte slices more compactly or more efficiently than other types of slices. 

On the other end I want to use [`deserialize_bytes`](https://docs.rs/serde/latest/serde/trait.Deserializer.html#tymethod.deserialize_bytes). I think this was working with Postcard because it happens to have the same representation, but I imagine other Serde impls could have different representations for "bytes" versus "seq of u8". We should probably test our code with something other than Postcard.

sffc replied:
Indeed, it breaks if I use `serialize_seq` with `deserialize_bytes` in the `bson` crate:

> called `Result::unwrap()` on an `Err` value: DeserializationError { message: "invalid type: sequence, expected a borrowed byte array" }

I would add a test case, except the `bson` crate pulls a bunch of new crates into the Cargo.lock file. To reproduce, add this to any of the test functions such as `test_serde_simpleascii_cow`:

```rust
let bson_bytes = bson::to_vec(&original).unwrap();
let bson_recovered: ZeroTrieSimpleAsciiCow = bson::from_slice(&bson_bytes).unwrap();
```

sffc replied:
I found another serde crate with the same behavior, `rmp-serde`, and added a test with it.

sffc replied:
Upstream Serde issue about `&[u8]` using serialize_seq instead of serialize_bytes:

https://github.com/serde-rs/serde/issues/2474

robertbastian commented on utils/zerotrie/src/options.rs:
```diff
@@ -8,44 +8,99 @@
 
+impl ZeroTrieBuilderOptions {
+    #[cfg(feature = "serde")]
+    pub(crate) const fn to_u8_flags(self) -> u8 {
```
nit: keep these `to_u8_flags` file-private, the rest of the crate should use `::OPTIONS`

sffc replied:
done

robertbastian reviewed (commented):
.

sffc commented:
I scope-creeped a little bit and added more `rmp-serde` tests. I eyeballed all the custom `Serialize` impls in the project and added some tests on ones that looked important or suspicious. I found and fixed 1 other instance where we were using `serialize_seq` with `deserialize_bytes`.

sffc commented:
Note: if you don't want me to serialize the whole thing as bytes, one potential alternative is that I could serialize a single `u8` followed by bytes.

sffc commented:
> Note: if you don't want me to serialize the whole thing as bytes, one potential alternative is that I could serialize a single `u8` followed by bytes.

^ Do you think doing it this way, with a separate entry encoding the discriminant, might be superior?

robertbastian commented:
Yes I think it would be.

sffc commented:
Ok coming right up

sffc commented:
ok I have it implemented, but there's a stable key `time_zone/iana_to_bcp47@1` that uses ZeroTrie and therefore the old format. I want to get rid of ZeroTrie and stick with just the variants (#4408) so I'm inclined to still use the new serialization format for the variants if we think it's superior.
